test/src/e2e.dynval/dv.m-m'.test.ts

```diff
@@ -0,0 +1,336 @@
+
+    const maxNumOfValidators = 6;
+    const alice = maxNumOfValidators - 1; // will be replaced
+    const bob = maxNumOfValidators; // will be elected by doing nothing
+    const charlie = maxNumOfValidators + 1; // will be elected by delegating enough
+    const dave = maxNumOfValidators + 2; // will be elected by depositing enough
+
+    const nodeParams = {
+        promiseExpect,
+        overrideParams: {
+            termSeconds,
+            maxNumOfValidators,
+            delegationThreshold: 1000,
+            minDeposit: 10000
+        },
+        validators: [
+            // Validators
+            { signer: validators[0], delegation: 5000, deposit: 100000 },
+            { signer: validators[1], delegation: 4900, deposit: 100000 },
+            { signer: validators[2], delegation: 4800, deposit: 100000 },
+            { signer: validators[3], delegation: 4700, deposit: 100000 },
+            { signer: validators[4], delegation: 4600, deposit: 100000 },
+            { signer: validators[5], delegation: 4000, deposit: 100000 }, // Alice
+            // Candidates
+            { signer: validators[6], delegation: 3000, deposit: 100000 }, // Bob
+            { signer: validators[7], delegation: 100, deposit: 100000 }, // Charlie
+            { signer: validators[8], delegation: 4100, deposit: 100 }, // Dave
+            { signer: validators[9], delegation: 100, deposit: 100 }
+        ]
+    };
+    const enoughDelegationToCatchBob = 3000;
+    const enoughDepositToCatchBob = 100000;
+    const aliceRevokeToBeLowerThanBob = 2000;
+    const enoughDelegationToCatchAlice = 4000;
+    const enoughDepositToCatchAlice = 100000;
+
+    async function expectAliceIsReplacedBy(
+        sdk: SDK,
+        name: string,
+        index: number
+    ) {
+        const possibleAuthors = await stake.getPossibleAuthors(sdk);
+        expect(possibleAuthors).not.to.be.null;
+        expect(possibleAuthors!.length).is.equals(maxNumOfValidators);
+        const authorAddresses = possibleAuthors!.map(x => x.toString());
+        expect(authorAddresses).contains.all.members(
+            validators.slice(0, alice).map(x => x.platformAddress.toString()),
+            "Contains previous validators except for Alice"
+        );
+        expect(authorAddresses).not.contains(
+            validators[alice].platformAddress.toString(),
+            "Alice should not be elected as a validator"
+        );
+        expect(authorAddresses).contains(
+            validators[index].platformAddress.toString(),
+            `${name} should be elected as a validator instead of alice`
+        );
+    }
+
+    describe("1. Jail one of the validator", async function() {
+        const nodes = withNodes(this, {
+            ...nodeParams,
+            onBeforeEnable: async bootstrappingNodes => {
+                await bootstrappingNodes[alice].clean(); // alice will be jailed!
+            }
+        });
+
+        beforeEach(async function() {
```

Comment:
It seems this block is repeated a lot.
How about moving these statements to the upper level `beforeEach`?